internal/local/structure_v2.go

```diff
@@ -292,33 +295,55 @@ func parseDataSources(rootDir string) ([]DataSourceStructure, error) {
 
 			colls := directoryWalker{path: dbPath, onlyDirs: true}
 			if err := colls.walk(func(coll os.FileInfo, collPath string) error {
+				// A valid data sources folder contains at least one of:
+				// - a rules.json file
+				// - a pair of files, schema.json and relationships.json
+				// If neither of these conditions are true (e.g. there is only a schema.json
+				// file, or there are no files in the directory at all), we should error
 
-				rulePath := filepath.Join(collPath, FileRules.String())
-				if _, err := os.Stat(rulePath); err != nil {
-					if os.IsNotExist(err) {
-						return nil // skip directories that do not contain `rules.json`
-					}
-					return err
-				}
+				// If we are not using app schemas, a valid data sources folder should contain all of
+				// these files and we should error otherwise
 
-				rule, err := parseJSON(rulePath)
+				rule, err := parseJSON(filepath.Join(collPath, FileRules.String()))
 				if err != nil {
 					return err
 				}
 
-				schema, err := parseJSON(filepath.Join(collPath, FileSchema.String()))
+				schemaBody, err := parseJSON(filepath.Join(collPath, FileSchema.String()))
 				if err != nil {
 					return err
 				}
-				rule[NameSchema] = schema
 
 				relationships, err := parseJSON(filepath.Join(collPath, FileRelationships.String()))
 				if err != nil {
 					return err
 				}
-				rule[NameRelationships] = relationships
 
-				rules = append(rules, rule)
+				if rule == nil && (schemaBody == nil || relationships == nil) {
```

Comment:
at the risk of extra indentation, we might be able to consolidate some of these conditionals

```go
if rule == nil {
  if schema == nil || relationships == nil { return fmt.Errorf(...) }
  schemas = append(schemas, map[string]interface{}{
    NameSchema: schema,
    NameRelationships: relationships,
    NameMetadata: ...,
  })
} else {
  if len(rule) == 0 { return errors.New(...) }
  rule[NameSchema] = schema
  rule[NameRelationshpis] = relationships,
}
```

[question] was it intentional that you dropped the `rule` assignments for schema and relationships when `rule != nil`?